This commit has brought contribution to streaming mini load

EmmyMiao87 opened the pull request:
* This commit has brought contribution to streaming mini load
The operation of streaming mini load is sames as previous. Also, user can check the load by frontend.
The difference is that streaming mini load finish the task before reply of REST API while the non-streaming only register a load.

* When updating doris
Updating fe or be firstly are also supported. After fe and be are updated, the streaming mini load will take effect.

* For multi mini load
The non-streaming mini load still has been used by multi mini load. The behavior of multi mini load has not been changed.

* Add unit test of broker load
Add unit test of broker load

EmmyMiao87 committed:
This commit has brought contribution to streaming mini load

* This commit has brought contribution to streaming mini load.
The operation of streaming mini load is sames as previous. Also, user can check the load by frontend.
The difference is that streaming mini load finish the task before reply of REST API while the non-streaming only register a load.

* When updating doris
Updating fe or be firstly are also supported. After fe and be are updated, the streaming mini load will take effect.

* For multi mini load
The non-streaming mini load still has been used by multi mini load. The behavior of multi mini load has not been changed.

* Add unit test of broker load
Add unit test of broker load

morningman commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +622,230 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+            if (hll_value.empty()) {
+                return Status("Hll value could not be empty when hll key is exists!"); 
+            }
+            std::map<std::string, std::string>* hll_map = new std::map<std::string, std::string>;
```
Memory leak.
Here you can just create the map on stack, not on heap.

EmmyMiao87 replied:
Fixed

morningman commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +622,230 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+    std::map<std::string, std::string> params(
+            req->query_params().begin(), req->query_params().end());
+    std::map<std::string, std::string>::iterator columns_it = params.find(COLUMNS_KEY);
+    if (columns_it != params.end()) {
```
Could you add a comment to this transform?
Using an example is better

EmmyMiao87 replied:
Added

morningman commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +622,230 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+
+void MiniLoadAction::_new_handle(HttpRequest* req) {
+    StreamLoadContext* ctx = (StreamLoadContext*) req->handler_ctx();
+    if (ctx == nullptr) {
```
Is it possible that we get a null StreamLoadContext here?
I think here should be an assertion(DCHECK(ctx != nullptr)), not an if condition.

EmmyMiao87 replied:
Changed

morningman commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +622,230 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+    std::string msg = "OK";
+    if (!ctx->status.ok()) {
+        // we do not send 500 reply to client, send 200 with error msg
+        status_str = "FAILED";
```
```suggestion
        status_str = "Failed";
```

EmmyMiao87 replied:
Should it be seems as previous?

morningman replied:
Um...Not professional, just change it~

EmmyMiao87 replied:
OK , I will change it

EmmyMiao87 committed:
Update the ut of MultiLoadMgr

morningman commented on fe/src/main/java/org/apache/doris/http/rest/GetLoadInfoAction.java:
```diff
@@ -63,7 +64,15 @@ public void executeWithoutPassword(AuthorizationInfo authInfo, BaseRequest reque
+            try {
+                catalog.getLoadInstance().getJobInfo(info);
+            } catch (DdlException e) {
+                catalog.getLoadManager().getLoadJobInfo(info);
```
I think you should first get info from your new Load Manager, then from the old Load Instance

EmmyMiao87 replied:
What‘s the difference between those two?

morningman replied:
LoadManager is your new implementation, so only history load job may exist in LoadInstance.
If you check LoadInstance first, then you have to check both LoadInstance and LoadManager for every load job info request in the future, which is inefficient.

EmmyMiao87 replied:
The hadoop load job still exists in the LoadInstance. In the future, it must check both LoadInstance and LoadManager for every load job info request. So the sequence is not important.

EmmyMiao87 committed:
Delete the plan ut

EmmyMiao87 committed:
Change if statement to DCHECK

EmmyMiao87 commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -443,6 +443,26 @@ struct TLoadCheckRequest {
+    9: optional i64 timeout_second
+    10: optional double max_filter_ratio 
+    11: optional i64 auth_code
+	12: required i64 create_timestamp;
```
Fixed

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadJob.java:
```diff
@@ -727,7 +773,7 @@ public void readFields(DataInput in) throws IOException {
         label = Text.readString(in);
         state = JobState.valueOf(Text.readString(in));
-        timeoutSecond = in.readInt();
+        timeoutSecond = in.readLong();
```
it isn't compatible with older version

EmmyMiao87 committed:
Change the doc of mini load

morningman reviewed (dismissed):
LGTM

imay commented on be/src/http/action/mini_load.cpp:
```diff
@@ -318,14 +330,47 @@ int MiniLoadAction::on_header(HttpRequest* req) {
+            [&res] (FrontendServiceConnection& client) {
+            client->isMiniLoadStreaming(res);
+            });
+    if (!status.ok()) {
```
I think you should check if res is ok. status ok means this RPC call was executed successfully, but the result may be  not ok

EmmyMiao87 replied:
The res in this RPC is not important, because the result of this RPC can only be successfully when fe has been finished updating.

imay replied:
this can be important now, but if we change the implementation, we can forget to change here.

And there is no harm to check res

EmmyMiao87 replied:
Em. I already changed.

imay commented on be/src/http/action/mini_load.h:
```diff
@@ -84,6 +88,27 @@ class MiniLoadAction : public HttpHandler {
+   
+    bool _is_streaming(HttpRequest* req);
+
+    bool isStreaming = true;
```
```suggestion
    bool _is_streaming = true;
```

imay commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +602,217 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+    std::map<std::string, std::string>::iterator columns_it = params.find(COLUMNS_KEY);
+    if (columns_it != params.end()) {
+        std::string columns_value = columns_it->second;
+        std::map<std::string, std::string>::iterator hll_it = params.find(HLL_KEY);
```
```suggestion
        auto hll_it = params.find(HLL_KEY);
```

imay commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +602,217 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+     * input: columns=c1,tmp_c2,tmp_c3\&hll=hll_c2,tmp_c2:hll_c3,tmp_c3
+     * output: columns=c1,tmp_c2,tmp_c3,hll_c2=hll_hash(tmp_c2),hll_c3=hll_hash(tmp_c3)
+     */
+    std::map<std::string, std::string>::iterator columns_it = params.find(COLUMNS_KEY);
```
```suggestion
    auto columns_it = params.find(COLUMNS_KEY);
```

imay commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +602,217 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+                return Status::InvalidArgument("Hll value could not tranform to hll expr: " + hll_value);
+            }
+            auto hll_map_it = hll_map.begin();
+            while (hll_map_it != hll_map.end()) {
```
I think what you want is `for (auto& it : hll_map)`

imay commented on be/src/http/action/mini_load.cpp:
```diff
@@ -516,4 +602,217 @@ bool LoadHandleCmp::operator() (const LoadHandle& lhs, const LoadHandle& rhs) co
+        }
+        put_request.__set_columns(columns_value);
+    }
+    std::map<std::string, std::string>::iterator column_separator_it = params.find(COLUMN_SEPARATOR_KEY);
```
```suggestion
    auto column_separator_it = params.find(COLUMN_SEPARATOR_KEY);
```

imay commented on be/src/runtime/stream_load/stream_load_executor.cpp:
```diff
@@ -223,17 +226,29 @@ void StreamLoadExecutor::rollback_txn(StreamLoadContext* ctx) {
 bool StreamLoadExecutor::collect_load_stat(StreamLoadContext* ctx, TTxnCommitAttachment* attach) {
-    if (ctx->load_type != TLoadType::ROUTINE_LOAD) {
+    if (ctx->load_type != TLoadType::ROUTINE_LOAD && ctx->load_type != TLoadType::MINI_LOAD) {
         // currently, only routine load need to set attachment
```
you should update comment too

imay commented on be/src/util/string_parser.hpp:
```diff
@@ -122,6 +123,38 @@ class StringParser {
+    template <typename T>
+    static Status split_string_to_map(const std::string& base, const T element_separator,
+                                   const T key_value_separator, 
+    							   std::map<std::string, std::string>* result) {
```
seems like there are some tabs. make them to spaces

imay commented on be/src/util/string_parser.hpp:
```diff
@@ -122,6 +123,38 @@ class StringParser {
+    static Status split_string_to_map(const std::string& base, const T element_separator,
+                                   const T key_value_separator, 
+    							   std::map<std::string, std::string>* result) {
+        if (result == nullptr) {
```
I think this is useless

EmmyMiao87 replied:
Maybe the input result is nullptr?

imay replied:
How can?

EmmyMiao87 replied:
It is my fault, I will change it.

imay commented on be/src/http/action/mini_load.cpp:
```diff
@@ -318,14 +330,47 @@ int MiniLoadAction::on_header(HttpRequest* req) {
+        std::stringstream ss;
+        ss << "This mini load is not streaming because: " << status.get_error_msg()
+		    << " with address(" << master_address.hostname << ":" << master_address.port << ")";
+        LOG(WARNING) << ss.str();
```
I think this is normal case when system is upgrading. so you should change WARNING to INFO

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -443,6 +443,26 @@ struct TLoadCheckRequest {
+    9: optional i64 timeout_second
+    10: optional double max_filter_ratio 
+    11: optional i64 auth_code
+    12: required i64 create_timestamp;
```
```suggestion
    12: optional i64 create_timestamp;
```

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -581,9 +608,14 @@ service FrontendService {
     TFeResult loadCheck(1:TLoadCheckRequest request)
+    // this method is used for streaming mini load
+    TMiniLoadBeginResult miniLoadBegin(TMiniLoadBeginRequest request)
+    TFeResult isMiniLoadStreaming()
```
I think you can make it more common function. like if we add some other functional, we can reuse this function.

So you can pass function in which is want to be checked

EmmyMiao87 replied:
OK， I will change it to another function

EmmyMiao87 committed:
Add a interface named isSupportedFunction

This function is used to protect the correctness of new feature which consists of be and fe during updaing.

EmmyMiao87 committed:
Persistence mini load after executed

EmmyMiao87 committed:
Fix compile error

EmmyMiao87 committed:
Fix the request of mini load

EmmyMiao87 committed:
Fix compile error

EmmyMiao87 committed:
Fix compile error

EmmyMiao87 committed:
Fix compile error

EmmyMiao87 committed:
Fix persistence error

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -443,6 +443,30 @@ struct TLoadCheckRequest {
+}
+
+struct TIsSupportedFunctionRequest {
+    1: required string function_name
```
And I think you can make it optional for extension

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/load/loadv2/MiniLoadJob.java:
```diff
@@ -0,0 +1,124 @@
+/*
```
Don't you think "/*" is more beautiful then "//"

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/common/FeMetaVersion.java:
```diff
@@ -116,4 +116,8 @@
+    // support strict mode and change timeout to long
+    public static final int VERSION_54 = 54;
+    // txn id has been record in load job
+    public static final int VERSION_55 = 55;
```
Because VERSION_54 already has been merged in master. Maybe someone already used in online environment.

EmmyMiao87 committed:
Change method name to isMethodSupported

EmmyMiao87 committed:
Change copyright of new file

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadJob.java:
```diff
@@ -1,22 +1,4 @@
- * under the License.
- *
- */
+
```
License header??

EmmyMiao87 committed:
Add copyright in load job
